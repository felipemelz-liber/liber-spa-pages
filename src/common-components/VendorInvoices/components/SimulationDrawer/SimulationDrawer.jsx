/* eslint-disable max-len */
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Icon from '@mdi/react';
import { mdiFinance } from '@mdi/js';
import { Drawer, Button } from 'liber-components';
import { Title, SelectedNumber, Row, ConfirmationButton } from './SimulationDrawer.styles';
import { SimulationInfo } from '../../../../views/fornecedor-duplicatas/components/SimulationDrawer/components/SimulationInfo/SimulationInfo';
import { SimulationDialog } from '../../../../views/fornecedor-duplicatas/components/SimulationDrawer/components/SimulationDialog/SimulationDialog';
import SimulationTerms from '../../../../views/fornecedor-duplicatas/components/SimulationDrawer/components/SimulationTerms/SimulationTerms';
import LoadingButton from '../../../Buttons/LoadingButton';
import {
  requestSimulationReport,
  anticipateInvoices,
  calculateQuote,
} from '../../actions/VendorInvoicesActions';
import { dispatchDataLayerEvent } from '../../../../vendor/Utils';

export const SimulationDrawer = ({
  selectedInvoices,
  quote,
  rateByPeriod,
  requestReport,
  requestAnticipation,
  changeQuote,
  loading,
  isSimulationOutdated,
}) => {
  React.useEffect(() => {
    if (selectedInvoices.length > 0) {
      changeQuote();
    }
  }, [selectedInvoices]);

  const [open, setOpen] = useState(false);
  const [canDownload, setCanDownload] = useState(false);

  const handleConfirmation = () => {
    dispatchDataLayerEvent('titulosBaixarRelatorioDaSimulacao');

    setOpen(false);
    setCanDownload(false);
    requestReport();
  };

  return (
    <>
      <Drawer
        blockNavigation={false}
        position="bottom"
        drawerSize={1120}
        size={128}
        show={selectedInvoices.length > 0}
      >
        <Row hasPaddingTop>
          <Title>
            Simulação de Adiantamento{' '}
            <SelectedNumber>{`(${selectedInvoices.length} títulos)`}</SelectedNumber>
          </Title>
          <Button version={2} size="small" onClick={() => setOpen(true)}>
            <Icon path={mdiFinance} />
            Baixar Relatório da Simulação (Excel)
          </Button>
        </Row>
        <Row>
          <SimulationInfo
            {...quote}
            selectedInvoices={selectedInvoices}
            rateByPeriod={rateByPeriod}
            isSimulationOutdated={isSimulationOutdated}
          >
            <LoadingButton
              onClick={requestAnticipation}
              loading={loading && !isSimulationOutdated}
              ButtonComponent={ConfirmationButton}
              size="large"
              disabled={isSimulationOutdated}
            >
              SOLICITAR ADIANTAMENTO
            </LoadingButton>
          </SimulationInfo>
        </Row>
      </Drawer>
      <SimulationDialog
        title="Termos do Relatório de Simulação"
        open={open}
        onClose={() => {
          setOpen(false);
          setCanDownload(false);
        }}
        maxWidth={592}
        confirmationButtonProps={{
          text: 'CONTINUAR',
          onConfirmation: handleConfirmation,
          disabled: !canDownload,
        }}
      >
        <SimulationTerms checked={canDownload} onChange={() => setCanDownload(!canDownload)} />
      </SimulationDialog>
    </>
  );
};

SimulationDrawer.propTypes = {
  selectedInvoices: PropTypes.arrayOf(PropTypes.number),
  quote: PropTypes.shape({
    faceValue: PropTypes.number,
    discountValue: PropTypes.number,
    discountRate: PropTypes.number,
    netValue: PropTypes.number,
  }),
  rateByPeriod: PropTypes.bool,
  requestReport: PropTypes.func,
  requestAnticipation: PropTypes.func,
  changeQuote: PropTypes.func,
  loading: PropTypes.bool,
  isAdmin: PropTypes.bool,
  isSimulationOutdated: PropTypes.bool,
};

SimulationDrawer.defaultProps = {
  selectedInvoices: [],
  quote: {},
  rateByPeriod: false,
  requestReport: () => {},
  requestAnticipation: () => {},
  changeQuote: () => {},
  loading: false,
  isAdmin: false,
  isSimulationOutdated: false,
};

const mapStateToProps = ({
  vendorInvoices: { selectedInvoices, quote, rateByPeriod, loading, isAdmin, isSimulationOutdated },
}) => ({
  selectedInvoices,
  quote,
  rateByPeriod,
  loading,
  isAdmin,
  isSimulationOutdated,
});

const mapDispatchToProps = {
  requestReport: requestSimulationReport,
  requestAnticipation: anticipateInvoices,
  changeQuote: calculateQuote,
};

export default connect(mapStateToProps, mapDispatchToProps)(SimulationDrawer);
