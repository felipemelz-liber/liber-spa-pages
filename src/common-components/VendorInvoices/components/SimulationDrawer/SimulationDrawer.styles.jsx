import styled, { css } from 'styled-components';
import { Button } from 'liber-components';

export const Title = styled.span`
  font-family: Roboto;
  font-size: 20px;
  font-weight: 500;
  font-stretch: normal;
  font-style: normal;
  line-height: 1.4;
  letter-spacing: normal;
  text-align: left;
  color: #4b6f85;
`;

export const SelectedNumber = styled(Title)`
  font-weight: normal;
`;

export const ConfirmationButton = styled(Button)`
  background-color: #002874;
  --button-ripple-color: #002d80;
  --button-text-color: #fafdff;
  min-width: 235px;
  &:hover {
    background-color: #001f59;
  }
`;

export const Row = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  padding: 16px;
  ${({ hasPaddingTop }) =>
    hasPaddingTop
      ? null
      : css`
          padding-top: 0px;
        `}
  align-items: center;
  & > div {
    margin-top: 0px;
  }
`;
