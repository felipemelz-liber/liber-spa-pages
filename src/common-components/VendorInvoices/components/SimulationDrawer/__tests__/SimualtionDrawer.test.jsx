import React from 'react';
import sinon from 'sinon';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { SimulationDrawer } from '../SimulationDrawer';
import LoadingButton from '../../../../../common-components/Buttons/LoadingButton';

describe('SimulationDrawer tests', () => {
  configure({ adapter: new Adapter() });

  const changeQuote = sinon.spy();
  const requestAnticipation = sinon.spy();

  const effectStub = sinon.stub(React, 'useEffect');

  const props = {
    selectedInvoices: [1, 2, 3],
    quote: {
      filed: 'mock',
    },
    rateByPeriod: true,
    changeQuote,
    requestAnticipation,
  };

  afterEach(() => {
    sinon.resetHistory();
  });

  afterAll(() => {
    effectStub.restore();
  });

  it('should render correctly with no props', () => {
    const wrapper = shallow(<SimulationDrawer />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render correctly with props', () => {
    const wrapper = shallow(<SimulationDrawer {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should call change quote correctly with selectedInvoices', () => {
    effectStub.callsArg(0);
    shallow(<SimulationDrawer {...props} />);

    expect(changeQuote.called).toBeTruthy();
  });

  it('should not call change quote without selectedInvoices', () => {
    effectStub.callsArg(0);
    shallow(<SimulationDrawer />);

    expect(changeQuote.called).toBeFalsy();
  });

  it('should call requestAnticipation on button click', () => {
    const wrapper = shallow(<SimulationDrawer {...props} />);

    const reportButton = wrapper.find(LoadingButton);
    reportButton.simulate('click');

    expect(requestAnticipation.called).toBeTruthy();
  });
});
