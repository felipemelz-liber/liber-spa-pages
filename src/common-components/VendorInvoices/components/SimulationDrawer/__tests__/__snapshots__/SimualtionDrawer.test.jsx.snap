// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimulationDrawer tests should render correctly with no props 1`] = `
<Fragment>
  <Drawer
    blockNavigation={false}
    closable={true}
    drawerSize={1120}
    duration={270}
    elementId="modal"
    fitScreen={false}
    onAppearStart={[Function]}
    onAppeared={[Function]}
    onLeaveStart={[Function]}
    onLeaved={[Function]}
    position="bottom"
    show={false}
    size={128}
  >
    <styled.div
      hasPaddingTop={true}
    >
      <styled.span>
        Simulação de Adiantamento
         
        <Styled(styled.span)>
          (0 títulos)
        </Styled(styled.span)>
      </styled.span>
      <Buttons__Button
        onClick={[Function]}
        size="small"
        version={2}
      >
        <Icon
          color="currentColor"
          horizontal={false}
          path="M6,16.5L3,19.44V11H6M11,14.66L9.43,13.32L8,14.64V7H11M16,13L13,16V3H16M18.81,12.81L17,11H22V16L20.21,14.21L13,21.36L9.53,18.34L5.75,22H3L9.47,15.66L13,18.64"
          rotate={0}
          size={null}
          spin={false}
          vertical={false}
        />
        Baixar Relatório da Simulação (Excel)
      </Buttons__Button>
    </styled.div>
    <styled.div>
      <SimulationInfo
        discountRate={0}
        discountValue={0}
        faceValue={0}
        inDialog={false}
        isSimulationOutdated={false}
        netValue={0}
        onRequestAnticipation={null}
        rateByPeriod={false}
        selectedInvoices={Array []}
      >
        <LoadingButton
          ButtonComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-htpNat",
                "isStatic": false,
                "rules": Array [
                  "
  background-color: #002874;
  --button-ripple-color: #002d80;
  --button-text-color: #fafdff;
  min-width: 235px;
  &:hover {
    background-color: #001f59;
  }
",
                ],
              },
              "displayName": "Styled(Buttons__Button)",
              "foldedComponentIds": Array [
                "Buttons__Button-u35aut-0",
              ],
              "render": [Function],
              "styledComponentId": "sc-htpNat",
              "target": Object {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          disabled={false}
          loading={false}
          onClick={[Function]}
          size="large"
          spinnerColor="white"
        >
          SOLICITAR ADIANTAMENTO
        </LoadingButton>
      </SimulationInfo>
    </styled.div>
  </Drawer>
  <SimulationDialog
    confirmationButtonProps={
      Object {
        "disabled": true,
        "onConfirmation": [Function],
        "text": "CONTINUAR",
      }
    }
    maxWidth={592}
    onClose={[Function]}
    open={false}
    title="Termos do Relatório de Simulação"
  >
    <SimulationTerms
      checked={false}
      onChange={[Function]}
    />
  </SimulationDialog>
</Fragment>
`;

exports[`SimulationDrawer tests should render correctly with props 1`] = `
<Fragment>
  <Drawer
    blockNavigation={false}
    closable={true}
    drawerSize={1120}
    duration={270}
    elementId="modal"
    fitScreen={false}
    onAppearStart={[Function]}
    onAppeared={[Function]}
    onLeaveStart={[Function]}
    onLeaved={[Function]}
    position="bottom"
    show={true}
    size={128}
  >
    <styled.div
      hasPaddingTop={true}
    >
      <styled.span>
        Simulação de Adiantamento
         
        <Styled(styled.span)>
          (3 títulos)
        </Styled(styled.span)>
      </styled.span>
      <Buttons__Button
        onClick={[Function]}
        size="small"
        version={2}
      >
        <Icon
          color="currentColor"
          horizontal={false}
          path="M6,16.5L3,19.44V11H6M11,14.66L9.43,13.32L8,14.64V7H11M16,13L13,16V3H16M18.81,12.81L17,11H22V16L20.21,14.21L13,21.36L9.53,18.34L5.75,22H3L9.47,15.66L13,18.64"
          rotate={0}
          size={null}
          spin={false}
          vertical={false}
        />
        Baixar Relatório da Simulação (Excel)
      </Buttons__Button>
    </styled.div>
    <styled.div>
      <SimulationInfo
        discountRate={0}
        discountValue={0}
        faceValue={0}
        filed="mock"
        inDialog={false}
        isSimulationOutdated={false}
        netValue={0}
        onRequestAnticipation={null}
        rateByPeriod={true}
        selectedInvoices={
          Array [
            1,
            2,
            3,
          ]
        }
      >
        <LoadingButton
          ButtonComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-htpNat",
                "isStatic": false,
                "rules": Array [
                  "
  background-color: #002874;
  --button-ripple-color: #002d80;
  --button-text-color: #fafdff;
  min-width: 235px;
  &:hover {
    background-color: #001f59;
  }
",
                ],
              },
              "displayName": "Styled(Buttons__Button)",
              "foldedComponentIds": Array [
                "Buttons__Button-u35aut-0",
              ],
              "render": [Function],
              "styledComponentId": "sc-htpNat",
              "target": Object {
                "$$typeof": Symbol(react.forward_ref),
                "render": [Function],
              },
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          disabled={false}
          loading={false}
          onClick={[Function]}
          size="large"
          spinnerColor="white"
        >
          SOLICITAR ADIANTAMENTO
        </LoadingButton>
      </SimulationInfo>
    </styled.div>
  </Drawer>
  <SimulationDialog
    confirmationButtonProps={
      Object {
        "disabled": true,
        "onConfirmation": [Function],
        "text": "CONTINUAR",
      }
    }
    maxWidth={592}
    onClose={[Function]}
    open={false}
    title="Termos do Relatório de Simulação"
  >
    <SimulationTerms
      checked={false}
      onChange={[Function]}
    />
  </SimulationDialog>
</Fragment>
`;
